Windows host process work

marosset opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature
/kind api-change


#### What this PR does / why we need it:
Adds alpha support for HostProcess containers on Windows

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Partially implements https://github.com/kubernetes/enhancements/issues/1981

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Add alpha support for HostProcess containers on Windows
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/1981-windows-privileged-container-support
```

/sig windows
/sig node
/milestone v1.21

marosset committed:
Adding WindowsHostProcessContainers feature flag

marosset committed:
Updating CRI API to include Windows Pod/Sandbox configs

marosset committed:
API support for Windows host process containers

Co-authored-by: James Sturtevant <jstur@microsoft.com>

marosset committed:
Kubelet updates for Windows HostProcess Containers

marosset commented:
/sig windows
/remove-sig apps
/triage accepted

marosset commented:
/label api-review

marosset commented:
/test pull-kubernetes-e2e-aks-engine-azure-windows
/test pull-kubernetes-e2e-aks-engine-windows-contianerd

marosset commented:
/assign @liggitt @yujuhong @Random-Liu

marosset committed:
Adding e2e test for HostProcess containers on Windows

marosset commented:
/retest

SergeyKanzhelev commented on test/e2e/windows/host_process.go:
```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
```suggestion
Copyright 2021 The Kubernetes Authors.
```

SergeyKanzhelev commented:
/priority important-soon

marosset commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5941,6 +5942,57 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+
+	podHostProcess := false
+	if podSpec.SecurityContext != nil && podSpec.SecurityContext.WindowsOptions != nil && podSpec.SecurityContext.WindowsOptions.HostProcess != nil {
+		podHostProcess = *podSpec.SecurityContext.WindowsOptions.HostProcess
```
I think it should - yes.
I am trying to follow to same inheritance logic as the other fields in WindowsSecurityContextOptions.
If the pod sets `RunAsUseraname : user1` but then all containers set `RunAsUsername: user2` then all containers would run as `user2`.

marosset replied:
Let me add a comment to add some clarity!

marosset replied:
Actually, I just checked against the KEP and we do want to fail validation if `pod hostProcess: false` but any containers set `hostProcess: true`

https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/1981-windows-privileged-container-support#kubernetes-api-updates

> If PodSecurityContext.WindowsSecurityContextOptions.HostProcess = false no containers may set Container.SecurityContext.WindowsSecurityContextOptions.HostProcess = true

SergeyKanzhelev replied:
is there a difference of explicit `false` and not set?

marosset replied:
Yes. If the value is not set at the pod-level we'll allow containers to be scheduled as `HostProcess` containers if all containers set the field.
If the value is set to false explicitly at the pod-level we'll fail validation.

Part of this behavior is forward-looking since we expect to be able to support pods with mixed `HostProcess` and `non-HostProcess` containers with some OS updates.
The other part of this behavior was to give cluster operators an easy way to author policies that forbid HostProcess containers. We see some practical use cases where HostProcess containers will be allowed in some pod namespaces like kube-system but forbidden from others.

marosset replied:
I'll add comments to clarify this

marosset replied:
Updated the validation logic in both kube-apiserver and kubelet and also added unit test cases in each to ensure if the Pod's WindowsOptions.HostProcess is explicitly set to false then the pod will not validate/schedule.

marosset commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5941,6 +5942,57 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+
+		// At present Windows Pods must container only HostProcess containers or regular containers.
+		if hostProcessContainerCount != containerCount {
+			errMsg := "If Pod contains any HostProcess containers then all containers must be HostProcess containers"
```
Good idea.
Service mesh inject is one of the scenarios we'd like to enable.
Let me see if I can re-work this.

marosset replied:
I'm finding it a bit difficult to do this cleanly.
If I can't get this done soon I will add a TODO in the code here to fix this before beta.
Would that be acceptable (since we are really close to code-freeze)?

marosset replied:
I updated the validation to list the names of containers that override HostProcess to be the opposite of what the Pod's HostProcess value is (which is disallowed). PTAL!

marosset commented on pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:
```diff
@@ -138,6 +140,14 @@ func (m *kubeGenericRuntimeManager) generatePodSandboxConfig(pod *v1.Pod, attemp
 	}
 	podSandboxConfig.Linux = lc
 
+	if runtime.GOOS == "windows" {
```
We want to set `CredentialSpec` and `RunAsUsername` even when the `WindowsHostProcessContainers` feature flag is not set.

For some time containerD will be looking for the fields under both LinuxOptions and WindowsOptions (preferring WindowsOptions if they exist). We're still working out the timelines for that tho...

marosset replied:
Actually this is more related to https://github.com/kubernetes/kubernetes/issues/92963 than #93580 but it will help with that one too. These CRI updates were discussed in the KEP https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/1981-windows-privileged-container-support#cri-implementation-details

marosset commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5941,6 +5942,57 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+		}
+	}
+
+	if hostProcessContainerCount > 0 {
```
I think that is the behavior we want. See https://github.com/kubernetes/kubernetes/pull/99576/files/760cbaf64d92a3b297337cffb70b29945cec71d0#r588565175

Thoughts?

marosset replied:
Hmm... I think it would be OK to schedule the Pod here. With other fields in WindowsOptions, if all containers set them to a different value that what was set in the Pod settings then what is specified in the Pod's WindowsOptions would be ignored.

For example if the Pod sets `PodSecurityContext.WindowsOptions.RunAsUsername=User1` but then all the containers set `SecurityContext.WindowsOptions.RunAsUsername=User2` then all of the Pods would run as 'User2`. 

Thoughts?

marosset replied:
I updated the validation logic to fail in this case and added a test case to the unit tests.

marosset replied:
The behavior where if `podHostProcess = true` then we should fail pod validation if any container sets `hostProcess = false` is described in the KEP under https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/1981-windows-privileged-container-support#windowssecuritycontextoptionshostprocess-flag 

If PodSecurityContext.WindowsSecurityContextOptions.HostProcess = false no containers may set Container.SecurityContext.WindowsSecurityContextOptions.HostProcess = true

(Sorry i missed this before)

marosset commented on pkg/apis/core/types.go:
```diff
@@ -5317,6 +5317,12 @@ type WindowsSecurityContextOptions struct {
 	RunAsUserName *string
+
+	// HostProcess determines if a container should be run as a 'Host Process' container.
+	// This field is alpha-level and will only be honored by components that enable the
```
I will update this comment to reflect how we will fail validation if flag is used by future is not set (like below)

marosset commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:
```diff
@@ -693,6 +695,29 @@ message LinuxContainerConfig {
+message WindowsSandboxSecurityContext {
+    // User name to run the container process as. If specified, the user MUST
+    // exist in the container image and be resolved there by the runtime;
+    // otherwise, the runtime MUST return error.
```
For this I just copied to comments from the fields that already existed on `WindowsContainerSecurityContext` .There are some open issues in k/k (https://github.com/kubernetes/kubernetes/issues/92963) where users are noticing that it is not possible to run the sandbox containers as a different user and this is possible on linux today.

I don't think we can fall back to the `run_as_user` - the types different (`run_as_user` is an Int64 and `run_as_username` is a string) and users are just handled different in linux vs windows.
I'm open to suggestions here.

marosset replied:
@jsturtevant I saw your below comment which is related 
> This does seem confusing to me. If I set a pod as podHostProcess=true I would expect all the containers to be hostProcess containers since we have the constraint that it is all or nothing. Is this for the usecase at somepoint when we may allow some containers to be hostprocess and others not?

Yea future support for pods container both hostprocess and non-hostprocess containers is the main reason why I didn't want o block pod validation if the pod sets HostProcess=true but all containers HostProcess=false.
Although this may be confusing I don't necessarily think we should block it.
I see this as more of a quirk when using setting inheritance.

Setting HostProcess=false at the Pod level was done to help out policy authors ensure HostProcess containers do not get scheduled if not allowed.

marosset replied:
I take back everything I just said...
The KEP discussions and KEP do state `If PodSecurityContext.WindowsSecurityContextOptions.HostProcess = true is set to true then no container in the pod sets Container.SecurityContext.WindowsSecurityContextOptions.HostProcess = false` 
Let me update this and add some additional test cases.
Sorry I mis-remembered this!

marosset commented on pkg/kubelet/container/helpers.go:
```diff
@@ -311,6 +312,34 @@ func HasPrivilegedContainer(pod *v1.Pod) bool {
 	return hasPrivileged
 }
 
+// HasHostProcessContainer returns true if any of the containers in a pod are HostProcess containers.
+func HasHostProcessContainer(pod *v1.Pod) bool {
```
updated

marosset commented:
/retest

erismaster commented:
Hi 👋 Bug Triage here, just a reminder that code freeze is end of day, tomorrow the 9th! PRs need to be merged in before then, or they will require an approved exception to go into the 1.21 milestone.

SergeyKanzhelev commented:
@marosset please ping when it's ready for the next review pass.

marosset commented:
@SergeyKanzhelev I think I addressed all of your feedback except listing out the individual container names that cause a pod to fail API validation. If you can take another look at everything else that would be great!

marosset commented:
Also, `pull-kubernetes-e2e-azure-disk-windows` is flakey and those failures there are not caused by updates in the PR (https://testgrid.k8s.io/sig-windows-presubmit#pr-k8s-azure-disk-e2e-master-windows)

jsturtevant commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -39,6 +40,8 @@ const (
 	containerPreStopHandlerLabel           = "io.kubernetes.container.preStopHandler"
 	containerPortsLabel                    = "io.kubernetes.container.ports"
+
+	windowsHostProcessContainer = "microsoft.com/hostprocess-container"
```
is formatting off here?

marosset replied:
Nope - I think because I separated this new label with a blank line (similar to lines able this snippet) it resets the justrification.

marosset replied:
Even before my changes the two conts starting with pod were has different spacing.

jsturtevant commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -90,7 +93,20 @@ func newPodLabels(pod *v1.Pod) map[string]string {
+	}
+
+	if kubecontainer.HasWindowsHostProcessContainer(pod) {
+		// While WindowsHostProcessContainers is in alpha pass 'io.microsoft.containers.hostProcess' annotation
```
'io.microsoft.containers.hostProcess' is not the value in the constant above `microsoft.com/hostprocess-container`

marosset replied:
fixed

jsturtevant commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -687,6 +689,30 @@ message LinuxContainerConfig {
+}
+
+    // Indicates whether the sandbox will be asked to run any HostProcess
+    // containers.
```
formatting looks off?

marosset replied:
fixed

SergeyKanzhelev commented on pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:
```diff
@@ -206,6 +216,56 @@ func (m *kubeGenericRuntimeManager) generatePodSandboxLinuxConfig(pod *v1.Pod) (
+	}
+
+	if sc.WindowsOptions != nil {
+		wo := sc.WindowsOptions
```
Do you mean: 

```suggestion
        if sc := pod.Spec.SecurityContext == nil || sc.WindowsOptions == nil {
            return wc, nil
        }
        wo := pod.Spec.SecurityContext.WindowsOptions
```

marosset replied:
I reformatted this block per Sergeys's comment below. We we sc.WindowsOptions enough below to warrant a new variable IMO

SergeyKanzhelev commented on pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:
```diff
@@ -206,6 +216,56 @@ func (m *kubeGenericRuntimeManager) generatePodSandboxLinuxConfig(pod *v1.Pod) (
+		SecurityContext: &runtimeapi.WindowsSandboxSecurityContext{},
+	}
+
+	sc := pod.Spec.SecurityContext
+	if pod.Spec.SecurityContext == nil {
+		return wc, nil
+	}
+
+	if sc.WindowsOptions != nil {
+		wo := sc.WindowsOptions
```
```suggestion
        if sc := pod.Spec.SecurityContext == nil || sc.WindowsOptions == nil {
            return wc, nil
        }
        wo := pod.Spec.SecurityContext.WindowsOptions
```

marosset replied:
Updated in the code. Thanks for the suggestion.

SergeyKanzhelev commented on test/e2e/windows/host_process.go:
```diff
@@ -0,0 +1,87 @@
+
+	ginkgo.It("should run as a process on the host/node", func() {
+
+		ginkgo.By("selecting a Windows node")
```
Add 
```
defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.WindowsHostProcessContainers, true)()
```

SergeyKanzhelev replied:
or some other way to validate that the feature gate was actually set for the test

marosset replied:
For the Windows E2E tests we manage the test configs at https://github.com/kubernetes-sigs/windows-testing/tree/master/job-templates and the test configs specify what feature gates are used.

We've got a separate issue for the Windows E2E tests where we don't have a good way to restrict test cases to only clusters with features enabled but we're working to solve that too.

SergeyKanzhelev replied:
you can try something similar to this: https://github.com/kubernetes/kubernetes/pull/99612/files#diff-471e8786474b22c41fa2d0ca2a9c3950a7da2e0ac7ab6de638c505204b39bf8dR142

SergeyKanzhelev commented on test/e2e/windows/host_process.go:
```diff
@@ -0,0 +1,87 @@
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+var _ = SIGDescribe("[Feature:WindowsHostProcessContainers] HostProcess containers", func() {
```
Also add `[Excluded:WindowsDocker]` similar to https://github.com/kubernetes/kubernetes/pull/97045/

marosset replied:
Added!

SergeyKanzhelev commented:
I do not see any major issues with this PR - a few minor fixes. So ping me for lgtm when updated. Let's try to merge before the code freeze tomorrow

marosset commented on pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:
```diff
@@ -206,6 +216,56 @@ func (m *kubeGenericRuntimeManager) generatePodSandboxLinuxConfig(pod *v1.Pod) (
+			}
+
+			if wo.HostProcess != nil && !*wo.HostProcess {
+				return nil, fmt.Errorf("Pod must not container any HostProcess containers if Pod's WindowsOptions.HostProcess is set to false")
```
fixed

marosset commented:
@SergeyKanzhelev @jsturtevant please take another look, I believe I addressed all of the feedback here.

I updated the API validation logic to report container names that cause the pod to fail validation and also addressed https://github.com/kubernetes/kubernetes/pull/99576#discussion_r589615948

marosset commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5976,6 +5977,71 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+	containerCount := 0
+	hostProcessContainerCount := 0
+
+	var podHostProcess *bool
```
This variable holds the value of the Pod's HostProcess value and doesn't indicate if the whole Pos Is HostProcess so I think it makes more sense to leave the `Is` out. Thoughts?

marosset commented:
Rebasing PR now...

marosset commented:
@liggitt Do you think you'd have a chance to review before code-freeze deadline for API changes?

SergeyKanzhelev commented on pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:
```diff
@@ -206,6 +216,54 @@ func (m *kubeGenericRuntimeManager) generatePodSandboxLinuxConfig(pod *v1.Pod) (
+			return nil, fmt.Errorf("Pod must not contain a both HostProcess and non-HostProcess containers")
+		}
+		if !kubecontainer.IsHostNetworkPod(pod) {
+			return nil, fmt.Errorf("HostNetwork is required if pod contains HostProcess containers")
```
super nit:

```suggestion
			return nil, fmt.Errorf("HostNetwork is required if Pod contains HostProcess containers")
```

Alternative - start all error messages above with lower case "p" in Pods to be consistent with GoLang error messages guidelines (error starts with lower case for better errors chaining)

marosset replied:
Let me do this in the next round of updates.

marosset replied:
updated

SergeyKanzhelev commented:
/lgtm

liggitt commented:
> Do you think you'd have a chance to review before code-freeze deadline for API changes?

this is on my list for tonight

yujuhong commented on pkg/features/kube_features.go:
```diff
@@ -723,6 +723,12 @@ const (
+	// owner: @marosset
+	// alpha: v1.21
+	//
+	// Enables scheduling of 'HostProcess' containers on Windows nodes
```
I think the feature supports more than simply "scheduling" such containers, right?

marosset replied:
Does `Enables support for 'HostProcess' containers on Windows nodes` sound better?

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6000,6 +6001,80 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+	}
+
+	// Note: Pod-wide WindowsSecurityContextOptions are inherited by containers that do not explicitly set them.
+	for _, c := range append(podSpec.Containers, podSpec.InitContainers...) {
```
use `VisitContainersWithPath(podSpec, fieldPath, func(container *api.Container, path *field.Path) bool { ... return true })` to visit all containers, rather than appending Containers and InitContainers

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6000,6 +6001,80 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+	containersWhichInheritHostProcessTrue := []string{}
+
+	var podHostProcess *bool
+	if podSpec.SecurityContext != nil && podSpec.SecurityContext.WindowsOptions != nil && podSpec.SecurityContext.WindowsOptions.HostProcess != nil {
```
can simplify to:
```suggestion
	if podSpec.SecurityContext != nil && podSpec.SecurityContext.WindowsOptions != nil {
```

jsturtevant replied:
i updated this

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6000,6 +6001,80 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+	}
+
+	hostNetwork := false
+	if podSpec.SecurityContext != nil && podSpec.SecurityContext.HostNetwork == true {
```
simplify to:
```suggestion
	if podSpec.SecurityContext != nil {
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6000,6 +6001,80 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+		if c.SecurityContext != nil && c.SecurityContext.WindowsOptions != nil && c.SecurityContext.WindowsOptions.HostProcess != nil {
+			if *c.SecurityContext.WindowsOptions.HostProcess {
+				containersWithHostProcessSetTrue = append(containersWithHostProcessSetTrue, c.Name)
+			} else if !*c.SecurityContext.WindowsOptions.HostProcess {
```
if it's non-nil, and isn't `true`, can simplify this to:
```suggestion
			} else {
```

marosset commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6000,6 +6001,80 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+		// At present Windows Pods which contain HostProcess containers must also set HostNetwork.
+		if hostNetwork != true {
+			errMsg := "PodSecurityContext.HostNetwork must be set if Pod contains any HostProcess containers"
+			allErrs = append(allErrs, field.Forbidden(fieldPath, errMsg))
```
In almost all of the failure cases here it is a combination of fields that leads to the error, not any specific field.
I couldn't come up with a consistent way to log errors for most of these cases so thought it would be better to set the pod spec as the field for most of these cases.

Do you have any suggestions on which field I should add the use set as triggering the error?
I didn't see any similar examples to follow here but am open to suggestions.

marosset replied:
I updated the section a bit and use fields in the error in some of the cases now. PTAL.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6000,6 +6001,80 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+			allErrs = append(allErrs, field.Forbidden(fieldPath, errMsg))
+		}
+
+		// If Pod's WindowsOptions.HostProcess is set to false, no container may set WindowsOptions.HostProcess to true.
+		if podHostProcess != nil && !*podHostProcess && len(containersWithHostProcessSetTrue) > 0 {
+			errMsg := fmt.Sprintf("Pod's WindowsOptions.HostProcess is set to false but the following containers override WindowsOptions.HostProcess to true: %v", containersWithHostProcessSetTrue)
+			allErrs = append(allErrs, field.Forbidden(fieldPath, errMsg))
+		}
```
would it be simpler to check this directly inside the container-visiting loop?

```go
for ... c in containers ...
  var containerHostProcess *bool
  if c.SecurityContext != nil && c.SecurityContext.WindowsOptions != nil {
    containerHostProcess := c.SecurityContext.WindowsOptions.HostProcess
  }

  if podHostProcess != nil && containerHostProcess != nil && *podHostProcess != *containerHostProcess {
    allErrs = append(allErrs, ... "pod hostProcess and container hostProcess must agree" error)
  }
}


```

marosset replied:
I was thinking of doing this but decided to group log all the errors at the end (mainly to make it more clear what all was being validated for the reader).

I think we'll still need some checks outside of the container-visiting loop but I'll can make some updates like you suggested to see if this simplifies things.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6000,6 +6001,80 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+		}
+	}
+
+	// If Pod's WindowsOptions.HostProcess is set to true, no contianer may set WindowsOptions.HostProcess to false.
+	if podHostProcess != nil && *podHostProcess && len(containersWithHostProcessSetFalse) > 0 {
+		errMsg := fmt.Sprintf("Pod's WindowsOptions.HostProcess is set to true but the following containers override WindowsOptions.HostProcess to false: %v", containersWithHostProcessSetFalse)
+		allErrs = append(allErrs, field.Forbidden(fieldPath, errMsg))
+	}
```
I think this would be covered by the `if podHostProcess != nil && containerHostProcess != nil && *podHostProcess != *containerHostProcess` check inside the container loop as well

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6000,6 +6001,80 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+
+	hostProcessContainerCount := len(containersWithHostProcessSetTrue) + len(containersWhichInheritHostProcessTrue)
+	if hostProcessContainerCount > 0 {
+		// Fail Pod validation if feature is not enabled instead of dropping fields based on PRR reivew.
```
> // Fail Pod validation if feature is not enabled instead of dropping fields based on PRR review.

Do you have a pointer to where this was discussed? The current validation code fails validation of existing persisted pods using `hostProcess:true` if the feature gate is disabled, preventing updates needed to do things like remove finalizers.

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version is the recommended approach to adding new alpha fields without breaking skewed API server validation.

marosset replied:
Part of the discussion happened here https://github.com/kubernetes/enhancements/pull/2288#discussion_r572301293 and the rest of the discussion happened on slack at https://kubernetes.slack.com/archives/CPNHUMN74/p1612805151060100.

Initially I was planning on dropping the fields like in the api_changes reference you linked but during the KEP review this was changed so as to not get into a situation where contianers that were meant to be scheduled as privileged/HostProcess containers would never get scheduled as regular containers if the fields got dropped.

liggitt replied:
hmm... that differs from the typical approach... I'm trying to think if that causes any issues

to fix the "invalidates persisted data" issue, rather than checking feature gates inside validation, add an `AllowWindowsHostProcessField bool` field to `PodValidationOptions` and populate it inside `GetValidationOptionsFromPodSpecAndMeta`:

```diff
diff --git a/pkg/api/pod/util.go b/pkg/api/pod/util.go
index 59f1a198e75..43c30d69751 100644
--- a/pkg/api/pod/util.go
+++ b/pkg/api/pod/util.go
@@ -402,6 +402,8 @@ func GetValidationOptionsFromPodSpecAndMeta(podSpec, oldPodSpec *api.PodSpec, po
 		AllowInvalidPodDeletionCost: !utilfeature.DefaultFeatureGate.Enabled(features.PodDeletionCost),
 		// Do not allow pod spec to use non-integer multiple of huge page unit size default
 		AllowIndivisibleHugePagesValues: false,
+		// Allow hostProcess field to be set if the feature is enabled
+		AllowWindowsHostProcessField: utilfeature.DefaultFeatureGate.Enabled(features.WindowsHostProcessContainers),
 	}
 
 	if oldPodSpec != nil {
@@ -419,6 +421,8 @@ func GetValidationOptionsFromPodSpecAndMeta(podSpec, oldPodSpec *api.PodSpec, po
 
 		// if old spec used non-integer multiple of huge page unit size, we must allow it
 		opts.AllowIndivisibleHugePagesValues = usesIndivisibleHugePagesValues(oldPodSpec)
+		// if old spec has hostProcess containers, allow them in the update request
+		opts.AllowWindowsHostProcessField = opts.AllowWindowsHostProcessField || setsHostProcessField(oldPodSpec)
 	}
 	if oldPodMeta != nil && !opts.AllowInvalidPodDeletionCost {
 		// This is an update, so validate only if the existing object was valid.
diff --git a/pkg/apis/core/validation/validation.go b/pkg/apis/core/validation/validation.go
index 1b668083b3b..5d4a88d6100 100644
--- a/pkg/apis/core/validation/validation.go
+++ b/pkg/apis/core/validation/validation.go
@@ -3204,6 +3204,8 @@ type PodValidationOptions struct {
 	AllowInvalidPodDeletionCost bool
 	// Allow pod spec to use non-integer multiple of huge page unit size
 	AllowIndivisibleHugePagesValues bool
+	// Allow hostProcess field to be set in the windows security context
+	AllowWindowsHostProcessField bool
 }
 
 // ValidatePodSingleHugePageResources checks if there are multiple huge
```

`setsHostProcessField` should be implemented to return true if the pod-level or any container-level `hostProcess` field is non-null

plumb PodValidationOptions to `validateWindowsHostProcessPod`, and disallow any non-nil `hostProcess` field value (true or false) if `AllowWindowsHostProcessField` is false

marosset replied:
Thanks! I'll try this out (but probably not tonight...)

marosset replied:
Alright I think I got all of this coded up. Thanks for the guidance here. This is new to me.

liggitt reviewed (commented):
I've reviewed the REST API and validation, but not the kubelet/node implementation and cri-api changes

marosset commented:
@liggitt @yujuhong  I think i addressed all of your feedback can two please take another look?
Thanks!

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6000,6 +6003,81 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+	}
+
+	podshelper.VisitContainersWithPath(podSpec, fieldPath, func(c *core.Container, cFieldPath *field.Path) bool {
+		containerCount++
```
separating the checks for "do container-level and pod-level hostProcess agree?" and "is this a hostProcess container?" would eliminate a lot of branching and make this shallower/easier-to-follow:

```go
if podHostProcess != nil && containerHostProcess != nil && *podHostProcess != *containerHostProcess {
	errMsg := fmt.Sprintf("pod and container windowsOptions.hostProcess must match if specified, pod was %v, container was %v", *podHostProcess, *containerHostProcess)
	allErrs = append(allErrs, field.Forbidden(cFieldPath.Child("securityContext","windowsOptions","hostProcess"), errMsg))
}

switch {
case containerHostProcess != nil && *containerHostProcess:
  // explicit container-level hostProcess=true
  hostProcessContainerCount++
case containerHostProcess == nil && podHostProcess != nil && *podHostProcess:
  // no container-level setting, explicit pod-level hostProcess=true
  hostProcessContainerCount++
}
```

dcantah commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:
```diff
@@ -693,6 +695,29 @@ message LinuxContainerConfig {
 
+// WindowsSandboxSecurityContext holds platform-specific configurations that will be
+// applied to a sandbox.
+message WindowsSandboxSecurityContext {
```
The settings will only apply to the pod sandbox itself so a comment like this would be warranted

marosset replied:
On Windows the `WindowsSandboxSecurityContext` will only apply to the sandbox container for a pod. I'm updating the comment to reflect that.

marosset commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6000,6 +6003,81 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+		// Fail Pod validation if feature is not enabled (unless podspec already exists and contains HostProcess fields) instead of dropping fields based on PRR reivew.
+		if !opts.AllowWindowsHostProcessField {
+			errMsg := "Pod contains Windows HostProcess containers but feature gate 'WindowsHostProcessContainers' is not enabled"
+			allErrs = append(allErrs, field.Forbidden(fieldPath, errMsg))
```
Whoops, missed this. 
Updated the logic and unit tests to fail if opts.AllowWindowsHostProcessField is not set and pod/containers set hostProcess (true or false)

marosset commented:
@yujuhong @liggitt please see my updates - thanks!

dchen1107 commented:
/approve

Please ping me once the rebase is done. Thanks!

dchen1107 commented:
/lgtm
/approve 

You might need API reviewer's approve here too

marosset commented:
Thanks @dchen1107 and @yujuhong!

Yes this needs API review/approval too.
I'm hoping @liggitt  can re-review before the deadline.

marosset commented:
/retest

annajung commented:
https://github.com/kubernetes/enhancements/issues/1981#issuecomment-794824236

/milestone clear

liggitt commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -687,6 +689,29 @@ message LinuxContainerConfig {
+// applied to a sandbox.
+// These settings will only apply to the sandbox container.
+message WindowsSandboxSecurityContext {
+    // User name to run the container process as. If specified, the user MUST
+    // exist in the container image and be resolved there by the runtime;
+    // otherwise, the runtime MUST return error.
+    string run_as_username = 1;
+
+    // The contents of the GMSA credential spec to use to run this container.
+    string credential_spec = 2;
```
this is surprising to me... was this missing from the v1 CRI API?

liggitt replied:
should this avoid mixing fixes or other additions into this PR, and keep this focused on the hostProcess addition?

marosset replied:
These updates were discussed in the hostProcess KEP but I can split them out into a separate commit or PR if that would be easier to track.

https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/1981-windows-privileged-container-support#cri-implementation-details

liggitt commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:
```diff
@@ -693,6 +695,29 @@ message LinuxContainerConfig {
+    // User name to run the container process as. If specified, the user MUST
+    // exist in the container image and be resolved there by the runtime;
+    // otherwise, the runtime MUST return error.
+    string run_as_username = 1;
+
+    // The contents of the GMSA credential spec to use to run this container.
+    string credential_spec = 2;
```
how were these fields being sent before?

dcantah replied:
They just weren't defined for the sandbox itself, they were present (besides `host_process` obviously) on the actual container config but not on this. https://github.com/kubernetes/kubernetes/pull/99576/files/87530d049994ec1aa113743bf1d20e434e422825#diff-583fcb2b2b9fed64e8ac45392565a924918b4a64a7d4b66df0d4bccd298e8769R722-R729

liggitt reviewed (commented):
I'm looking primarily at the REST API and validation changes, it looks like the kubelet and cri-api packages already got review (though I did have a question about mixing other CRI-API changes into this PR)

The field definition, docs, and validation all look as expected and lgtm. My one question is about whether we should hard-fail or drop hostProcess fields from incoming pods if the feature is not enabled.

hoisting from https://github.com/kubernetes/kubernetes/pull/99576#discussion_r589954739

@marosset:
> // Fail Pod validation if feature is not enabled instead of dropping fields based on PRR review.

@liggitt:
> Do you have a pointer to where this was discussed? The current validation code fails validation of existing persisted pods using `hostProcess:true` if the feature gate is disabled, preventing updates needed to do things like remove finalizers.
> 
> https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version is the recommended approach to adding new alpha fields without breaking skewed API server validation.

@marosset:
> Part of the discussion happened here kubernetes/enhancements#2288 (comment) and the rest of the discussion happened on slack at https://kubernetes.slack.com/archives/CPNHUMN74/p1612805151060100.
> 
> Initially I was planning on dropping the fields like in the api_changes reference you linked but during the KEP review this was changed so as to not get into a situation where contianers that were meant to be scheduled as privileged/HostProcess containers would never get scheduled as regular containers if the fields got dropped.

From the linked slack conversation:

@deads2k:
> at some point, the user creating the pod needs to know "this isn't going to work as I expected it to"
>
> in a perfect world, this is done with a 400series validation error
>
> in a very imperfect world, they do a POST/GET cycle and see their field isn't persisted so it won't work (edited) 
>
> in a slightly easier world, you provide them with a snipped to put in a script to ensure they're running privileged so their pod doesn't do something  unexepcted

Given <= 1.20 servers don't have this field, wouldn't clients have to do a POST/GET to know if their hostProcess field was persisted or not anyway? Those servers don't fail requests if a hostProcess field is set, they just drop it. I'm not sure a server with this field in the go type but which has the feature disabled should behave different.y

marosset commented:
/milestone v1.22

marosset commented:
/retest

marosset commented:
@liggitt @deads2k 
re: 
> My one question is about whether we should hard-fail or drop hostProcess fields from incoming pods if the feature is not enabled.

and 

> Given <= 1.20 servers don't have this field, wouldn't clients have to do a POST/GET to know if their hostProcess field was persisted or not anyway? Those servers don't fail requests if a hostProcess field is set, they just drop it. I'm not sure a server with this field in the go type but which has the feature disabled should behave differently

I don't have a have a strong opinion on if we should drop the fields or hard-fail the pod.
My _preference_ here would be to fail the pod since I think that would provide a better user experience for people experimenting with this alpha feature (as long as we can do so safely and not leave zombie pods around - which should be the case here)

marosset commented:
> @liggitt @deads2k
> re:
> 
> > My one question is about whether we should hard-fail or drop hostProcess fields from incoming pods if the feature is not enabled.
> 
> and
> 
> > Given <= 1.20 servers don't have this field, wouldn't clients have to do a POST/GET to know if their hostProcess field was persisted or not anyway? Those servers don't fail requests if a hostProcess field is set, they just drop it. I'm not sure a server with this field in the go type but which has the feature disabled should behave differently
> 
> I don't have a have a strong opinion on if we should drop the fields or hard-fail the pod.
> My _preference_ here would be to fail the pod since I think that would provide a better user experience for people experimenting with this alpha feature (as long as we can do so safely and not leave zombie pods around - which should be the case here)

This was discussed on slack and the conclusion was to keep rejection in validation and not drop fields.
https://kubernetes.slack.com/archives/CPNHUMN74/p1619719023028700

liggitt commented on pkg/apis/core/types.go:
```diff
@@ -5350,6 +5350,13 @@ type WindowsSecurityContextOptions struct {
+	// HostProcess determines if a container should be run as a 'Host Process' container.
+	// This field is alpha-level and will only be honored by components that enable the
+	// WindowsHostProcessContainers feature flag. Setting this field without the feature
+	// flag will result errors when validating the Pod.
```
This should also describe the requirement that all containers effective `hostProcess` value must agree, and that if a pod is `hostProcess`, it must also set `hostNetwork` to true

jsturtevant replied:
I've updated the wording to reflect this requirement

jsturtevant commented:
Not sure why prow reported all those tests failed. Github only reported gce alpha 

/retest

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5974,6 +5977,91 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+
+		if !opts.AllowWindowsHostProcessField && containerHostProcess != nil {
+			// Do not allow pods to persist data that sets hostProcess (true or false)
+			errMsg := "container must not specify securityContext.WindowsOptions.hostProces when feature gate 'WindowsHostProcessContainers' is not enabled"
```
the field path is already prefixed to the error, just explain why the value is not permitted:

```suggestion
			errMsg := "not allowed when feature gate 'WindowsHostProcessContainers' is not enabled"
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5974,6 +5977,91 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+
+	if !opts.AllowWindowsHostProcessField && podHostProcess != nil {
+		// Do not allow pods to persist data that sets hostProcess (true or false)
+		errMsg := "pod must not specify securityContext.WindowsOptions.hostProces when feature gate 'WindowsHostProcessContainers' is not enabled"
```
the field path is already prefixed to the error, just explain why the value is not permitted:


```suggestion
		errMsg := "not allowed when feature gate 'WindowsHostProcessContainers' is not enabled"
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5974,6 +5977,91 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+		}
+
+		if podHostProcess != nil && containerHostProcess != nil && *podHostProcess != *containerHostProcess {
+			errMsg := fmt.Sprintf("pod and container windowsOptions.hostProcess must match if specified, pod was %v, container was %v", *podHostProcess, *containerHostProcess)
+			allErrs = append(allErrs, field.Forbidden(cFieldPath.Child("securityContext", "windowsOptions", "hostProcess"), errMsg))
```
```suggestion
			errMsg := fmt.Sprintf("pod hostProcess value must be identical if both are specified, was %v", *podHostProcess)
			allErrs = append(allErrs, field.Invalid(cFieldPath.Child("securityContext", "windowsOptions", "hostProcess"), *containerHostProcess, errMsg))
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5974,6 +5977,91 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+
+		// At present Windows Pods which contain HostProcess containers must also set HostNetwork.
+		if hostNetwork != true {
+			errMsg := "hostNetwork must be set if pod contains any hostProcess containers"
+			allErrs = append(allErrs, field.Forbidden(fieldPath.Child("hostNetwork"), errMsg))
```
```suggestion
			errMsg := "hostNetwork must be true if pod contains any hostProcess containers"
			allErrs = append(allErrs, field.Invalid(fieldPath.Child("hostNetwork"), hostNetwork, errMsg))
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5974,6 +5977,91 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
+		// HostProcess containers (explicitly set or inherited).
+		if hostProcessContainerCount != containerCount {
+			errMsg := "If pod contains any hostProcess containers then all containers must be HostProcess containers"
+			allErrs = append(allErrs, field.Forbidden(fieldPath, errMsg))
```
```suggestion
			allErrs = append(allErrs, field.Invalid(fieldPath, "", errMsg))
```

liggitt commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -38,6 +39,8 @@ const (
 	containerPreStopHandlerLabel           = "io.kubernetes.container.preStopHandler"
 	containerPortsLabel                    = "io.kubernetes.container.ports"
+
+	windowsHostProcessContainer = "microsoft.com/hostprocess-container"
```
a vendor label domain here is surprising... is this the source of this label or is this driving existing behavior elsewhere?

liggitt replied:
the comments indicate this is only expected while the feature is in alpha? what's the plan to transition this from an annotation to a field?

jsturtevant replied:
> a vendor label domain here is surprising... is this the source of this label or is this driving existing behavior elsewhere?

The label is used by hcsshim: https://github.com/microsoft/hcsshim/blob/407147a8825c21b2f1f1ef4e53b24b27fb8632fb/internal/oci/uvm.go#L94  

@dcantah might be able to shed more light on the source

> the comments indicate this is only expected while the feature is in alpha? what's the plan to transition this from an annotation to a field?

Yes the plan is to integrate this directly into containerd instead of using the annotation.

dcantah replied:
I'm not really sure if this is the right place for the annotation/to be setting it. For some background, the plan was if everything went smoothly and we were able to get a mirroring field on the OCI runtime spec for this, to translate the field added in this PR to the new OCI spec field. If the OCI changes fall through, ideally we'd check if the HostProcess field is set on the cri spec and set this annotation in the cri plugin/containerd. 

I think we can probably remove this

dcantah replied:
Although for Alpha just in case there's any challenges/pushback with adding a change to set these annotations in the cri plugin (I can't see how but just to be safe) I think it's alright to keep this, but obviously I'm not the last say here.

liggitt replied:
> The label is used by hcsshim: https://github.com/microsoft/hcsshim/blob/407147a8825c21b2f1f1ef4e53b24b27fb8632fb/internal/oci/uvm.go#L94

Is the kubelet wired directly to hcsshim on windows? if not, setting a label specifically for that runtime is weirdly coupling

> > what's the plan to transition this from an annotation to a field?
> 
> Yes the plan is to integrate this directly into containerd instead of using the annotation.

I more meant what protects against this code persisting beyond the alpha phase?

jsturtevant replied:
> Is the kubelet wired directly to hcsshim on windows? 

Not for containerd, dockershim uses hcsshim directly in some cases but not for starting containers.

> I more meant what protects against this code persisting beyond the alpha phase?

The plan is to remove it when moving to beta.  I don't know exactly the best way to protect against persistence.  Would making it a requirement to remove it for graduation to Beta work?

Otherwise I agree this is a strange coupling and we could remove the annotations here.  This will require updates to containerd to parse the CRI fields before it could be used (where as with the annotations it is usable with no changes to containerd). Containerd will read from the CRI field and move the value to  the OCI annotations while we are waiting for https://github.com/opencontainers/runtime-spec/pull/1098 to be accepted in the OCI spec.

liggitt replied:
> Would making it a requirement to remove it for graduation to Beta work?

yeah, that seems reasonable

jsturtevant replied:
I've updated the KEP to reflect this: https://github.com/kubernetes/enhancements/pull/2755

liggitt commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -89,7 +92,20 @@ func newPodLabels(pod *v1.Pod) map[string]string {
 
 // newPodAnnotations creates pod annotations from v1.Pod.
 func newPodAnnotations(pod *v1.Pod) map[string]string {
-	return pod.Annotations
```
should this keep the existing behavior if the os is not windows or if the pod has no hostprocess containers?

jsturtevant replied:
It looks like the result would be the same, but there is an added iteration over the annotations.  the `pod.Annotations` are copied to `annotations` and only if a hostprocesscontainer is present then an additional annotation is applied.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5974,6 +5977,91 @@ func validateWindowsSecurityContextOptions(windowsOptions *core.WindowsSecurityC
 	return allErrs
 }
 
+func validateWindowsHostProcessPod(podSpec *core.PodSpec, fieldPath *field.Path, opts PodValidationOptions) field.ErrorList {
```
I don't think addition of ephemeral containers to a pod go through this validation path. Before this or ephemeral containers graduates from alpha, we need to ensure ephemeralcontainers are also checked for enablement of hostProcess and consistency of effective hostProcess.

cc @verb

can both of you add a tracking item to your alpha graduation criteria so this doesn't get dropped?

verb replied:
Interesting. Agreed, it would be safer to validate the entire PodSpec after updating the ephemeral containers.

jsturtevant replied:
https://github.com/kubernetes/enhancements/pull/2748

liggitt commented:
alpha job failure is https://github.com/kubernetes/test-infra/pull/22035/files#r632593674

jsturtevant commented:
The aks-engine failure is being fixed via https://github.com/Azure/aks-engine/pull/4412

jsturtevant commented:
aks-engine tests should be fixed now.  
/retest

jsturtevant commented:
>I did not review the kubelet and CRI functionality and API otherwise, still needs node/windows review

For Windows
/assign @jayunit100 @claudiubelu 

For Node
/assign @SergeyKanzhelev @yujuhong 

fyi the `pull-kubernetes-e2e-gce-alpha-features` test still looks like it is failing. Looks like https://github.com/kubernetes/kubernetes/pull/100136

jayunit100 commented:
code looks good - any instructions for smoke testing it that i can follow ?

jayunit100 commented:
/retest

jayunit100 commented:
/lgtm

jayunit100 commented:
imo this is alpha , from windows perspective, lets merge and iterate !

SergeyKanzhelev commented on api/openapi-spec/swagger.json:
```diff
@@ -10989,6 +10989,10 @@
           "type": "string"
         },
+        "hostProcess": {
+          "description": "HostProcess determines if a container should be run as a 'Host Process' container. This field is alpha-level and will only be honored by components that enable the WindowsHostProcessContainers feature flag. Setting this field without the feature flag will result errors when validating the Pod.",
```
```suggestion
          "description": "HostProcess determines if a container should be run as a 'Host Process' container. This field is alpha-level and will only be honored by components that enable the WindowsHostProcessContainers feature flag. Setting this field without the feature flag will result in errors when validating the Pod.",
```

jsturtevant replied:
Is this a generated file?  I thought if I updated the other values it would it would fix this. I ran `make clean && make generated_files` based on https://github.com/kubernetes/community/blob/4026287dc3a2d16762353b62ca2fe4b80682960a/contributors/devel/sig-architecture/api_changes.md#edit-version-conversions

SergeyKanzhelev replied:
I typically run `make update`, but the same idea. Should be auto-generated.

jsturtevant replied:
make update did the trick

liggitt commented:
cri-api changes approved by dchen1107,yujuhong

/approve

SergeyKanzhelev commented on pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:
```diff
@@ -206,6 +216,54 @@ func (m *kubeGenericRuntimeManager) generatePodSandboxLinuxConfig(pod *v1.Pod) (
+		// enabled instead of trying to schedule containers as regular containers as stated in
+		// PRR review.
+		if !utilfeature.DefaultFeatureGate.Enabled(features.WindowsHostProcessContainers) {
+			return nil, fmt.Errorf("pod contains HostProcess containeers but feature 'WindowsHostProcessContainers' is not enabled")
```
```suggestion
			return nil, fmt.Errorf("pod contains HostProcess containers but feature 'WindowsHostProcessContainers' is not enabled")
```

SergeyKanzhelev replied:
same in corresponding test

SergeyKanzhelev commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -38,6 +39,8 @@ const (
 	containerPreStopHandlerLabel           = "io.kubernetes.container.preStopHandler"
 	containerPortsLabel                    = "io.kubernetes.container.ports"
+
+	windowsHostProcessContainer = "microsoft.com/hostprocess-container"
```
I see the discussion that this is temporary label. I think the comment in code will help to highlight this.

jsturtevant replied:
added comment

SergeyKanzhelev commented on pkg/kubelet/kuberuntime/labels.go:
```diff
@@ -89,7 +92,20 @@ func newPodLabels(pod *v1.Pod) map[string]string {
+		annotations[k] = v
+	}
+
+	if kubecontainer.HasWindowsHostProcessContainer(pod) {
```
should this logic only execute in Windows build of kubelet?

SergeyKanzhelev replied:
same qq - whether it should run under feature flag only?

jsturtevant replied:
This seems like a reasonable request.  The generate `generatePodSandboxWindowsConfig` is behind a GOOS flag and there is a feature flag check there as well. 

looks like we might want it https://github.com/kubernetes/kubernetes/pull/99576/files#diff-49d1c9b046a0055e2c7cd93a85f90131819eb5e30eaaac1daf3ad66a289e5af3R125-R126 and https://github.com/kubernetes/kubernetes/pull/99576/files#diff-49d1c9b046a0055e2c7cd93a85f90131819eb5e30eaaac1daf3ad66a289e5af3R125-R126 as well

The pod its self would be rejected due to https://github.com/kubernetes/kubernetes/pull/99576/files#diff-2e2d495cb2f6938efb55dfcd758c182922b2f28271f9c237bf710c49cb8a12a7R245 so those annotations would never make it through to actual create request to the runtime.

What do you think?

SergeyKanzhelev replied:
all linked examples are pointing to either files with `_windows.go` or called under GOOS. I understand this code will never result in setting the label. It also will be removed in beta. This comment is just to be on a safe side and only call this under GOOS or with the feature flag. This is what I typically pay attention when review PRs introducing the new feature flags.

jsturtevant replied:
Will make those updates

jsturtevant replied:
I put this on a separate commit for easier review: https://github.com/kubernetes/kubernetes/pull/99576/commits/9fc16f4ed58f9ed2453e5b81216a36292686fe25

If it looks good I can rebase it into the other commits.

SergeyKanzhelev commented on staging/src/k8s.io/api/core/v1/generated.proto:
```diff
@@ -5629,5 +5629,12 @@ message WindowsSecurityContextOptions {
+  // HostProcess determines if a container should be run as a 'Host Process' container.
+  // This field is alpha-level and will only be honored by components that enable the
+  // WindowsHostProcessContainers feature flag. Setting this field without the feature
+  // flag will result errors when validating the Pod.
```
```suggestion
  // flag will result in errors when validating the Pod.
```

SergeyKanzhelev commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -6215,6 +6215,13 @@ type WindowsSecurityContextOptions struct {
+	// HostProcess determines if a container should be run as a 'Host Process' container.
+	// This field is alpha-level and will only be honored by components that enable the
+	// WindowsHostProcessContainers feature flag. Setting this field without the feature
+	// flag will result errors when validating the Pod.
```
```suggestion
	// flag will result in errors when validating the Pod.
```

finally commented on the source of it =)

SergeyKanzhelev commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -687,6 +689,29 @@ message LinuxContainerConfig {
+    // The contents of the GMSA credential spec to use to run this container.
+    string credential_spec = 2;
+
+    // Indicates whether the container be asked to run as a HostProcess container.
```
strange "be asked". Use "requested"? Or even present tense: "requests"?

```suggestion
    // Indicates whether the container requested to run as a HostProcess container.
```

SergeyKanzhelev commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -687,6 +689,29 @@ message LinuxContainerConfig {
+    // otherwise, the runtime MUST return error.
+    string run_as_username = 1;
+
+    // The contents of the GMSA credential spec to use to run this container.
```
nit: is this a json "as is"? Sometimes there is a confusion when json used whether wrapping braces are included or not.

jsturtevant replied:
This was copied over from the ContainerSecurityContext: https://github.com/kubernetes/kubernetes/blob/bc637a7a07cf917140daa3bd434eac9bbd8fe92c/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto#L697-L704 

The json used here is passed directly through to the runtime as is and I believe it includes the brakets.  The way this is used in practice is via the gmsa webhook: https://github.com/kubernetes-sigs/windows-gmsa/ which converts a CRD to the json.

SergeyKanzhelev replied:
resolved

SergeyKanzhelev commented on test/e2e/windows/host_process.go:
```diff
@@ -0,0 +1,96 @@
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+var _ = SIGDescribe("[Feature:WindowsHostProcessContainers] [Excluded:WindowsDocker] HostProcess containers", func() {
```
```suggestion
var _ = SIGDescribe("[Feature:WindowsHostProcessContainers] [Excluded:WindowsDocker] [MinimumKubeletVersion:1.22] HostProcess containers", func() {
```

liggitt commented:
/hold for a few node comments

SergeyKanzhelev commented:
> /assign @SergeyKanzhelev @yujuhong

added a few comments. Nothing blocking. Will lgtm when comments are looked at.

SergeyKanzhelev commented:
/lgtm

swagger comment is a minor issue. If build ok, pr is ok.

jsturtevant commented:
> swagger comment is a minor issue. If build ok, pr is ok.

Verify failed because swagger was out of date.  I've fixed that

SergeyKanzhelev commented:
/lgtm

jsturtevant commented:
/retest

jsturtevant commented:
Are we ready to remove the hold?  I believe the gce-alpha job is not required and is a fix is being tracked in  #100136

liggitt commented:
/hold cancel
